autopush/web/webpush.py

```diff
@@ -56,13 +56,21 @@ def post(self, api_ver="v1", token=None):
     def _router_completed(self, response, uaid_data, warning=""):
         """Called after router has completed successfully"""
         # Were we told to update the router data?
-        # GCM/APNS bridges can result in data updates
         if response.router_data is not None:
             if not response.router_data:
-                del uaid_data["router_data"]
-            else:
-                uaid_data["router_data"] = response.router_data
+                # An empty router_data object indicates that the record should
+                # be deleted. There is no longer valid route information for
+                # this record.
+                self.ap_settings.router.drop_user(self.uaid)
+                return self._router_response(response)
+            # The router data needs to be updated to include any changes
+            # requested by the bridge system
+            uaid_data["router_data"] = response.router_data
+            # set the AWS mandatory data
             uaid_data["connected_at"] = ms_time()
+            rt = self.valid_input.get(
```

Comment:
Also, the valid_input was guaranteed to have a router_type and subscription attribute already, the benefit of validation up front means we aren't guessing if it exists and using get calls like this.